File: windows/WallpaperInfoApp/WallpaperInfoApp/WallpaperInfoUI.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Windows.Forms.Integration;
using System.Globalization;
using System.Text.RegularExpressions;
using WpfCustomControlLib;

namespace WallpaperInfoApp {

    public partial class WallpaperInfoUI : Form {

        public WallpaperInfoUI() {
            InitializeComponent();
			// Note : Label doesn't support DPI awareness well so use TextBox with readonly option instead. ("Theme", "Custom")
        }

		private void windowLoad(object sender, EventArgs e) {
			if (AppUtil.isMyServiceRunning()) {
				_serviceRunning = true;
				_serviceConnection = WallpaperServiceConnection.bindService(this);
			} else {
				_serviceRunning = false;
			}
			_themeComboBox.Items.AddRange(ThemeInfo.getLabels().ToArray());
			updateLayoutWithServiceInfo();
		}

		private void chooseDirButtonClicked(object sender, EventArgs e) {
            BPUtil.BPLog("WallpaperInfoUI.chooseDirButtonClicked");
			WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
			var path = DirectoryChooserDialog.directoryChooserDialog(this, wpsi.getSourceRootPath());
			if (path != null) {
				_serviceConnection.sendMessageToService(command: MSG.SET_ROOT, objectOption: path);
				updateLayoutWithServiceInfo();
			}
		}

		private void themeComboBoxChanged(object sender, EventArgs e) {
            BPUtil.BPLog("WallpaperInfoUI.themeComboBoxChanged");
			if (_doneLoading) {
				var index = _themeComboBox.SelectedIndex;
				if (_serviceConnection != null) {
					_serviceConnection.sendMessageToService(command: MSG.SET_THEME, intOption: index);
				}			
			}
		}

		private bool _trackbarMouseDown = false;
		private bool _trackbarScrolling = false;

		private void intervalTrackBar_MouseDown(object sender, MouseEventArgs e) {
			_trackbarMouseDown = true;
		}

		private void intervalTrackBar_MouseUp(object sender, MouseEventArgs e) {
			if (_trackbarMouseDown == true && _trackbarScrolling == true) {
				if (_serviceRunning) {
					_serviceConnection.sendMessageToService(command: MSG.SET_INTERVAL, intOption: _intervalTrackBar.Value);
				}
			}
			_trackbarMouseDown = false;
			_trackbarScrolling = false;
		}

		private void intervalTrackBar_MouseMove(object sender, MouseEventArgs e) {
			_trackbarScrolling = true;
		}

		private void startStopButtonClicked(object sender, EventArgs e) {
            BPUtil.BPLog("WallpaperInfoUI.startStopButtonClicked");
			if (AppUtil.isMyServiceRunning()) {
				_serviceRunning = true;
			} else {
				_serviceRunning = false;
			}
        
			if (_serviceRunning) {
				WallpaperServiceConnection.stopService();
				_serviceConnection.unbindService();
				_serviceConnection = null;
			} else {
				_serviceConnection = WallpaperServiceConnection.bindService(wallpaperInfoUI: this);
				WallpaperServiceConnection.startService();
			}

			_serviceRunning = !_serviceRunning;
			updateLayoutWithServiceInfo();
		}

		private void previousButtonClicked(object sender, EventArgs e) {
            BPUtil.BPLog("WallpaperInfoUI.previousButtonClicked");
			_serviceConnection.sendMessageToService(command: MSG.PREVIOUS);
		}

		private void pauseButtonClicked(object sender, EventArgs e) {
            BPUtil.BPLog("WallpaperInfoUI.pauseButtonClicked");
			WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
			_serviceConnection.sendMessageToService(command: MSG.PAUSE, intOption: (!wpsi.getPause()) ? 1 : 0);
			updateLayoutWithServiceInfo();
		}

		private void nextButtonClicked(object sender, EventArgs e) {
            BPUtil.BPLog("WallpaperInfoUI.nextButtonClicked");
			_serviceConnection.sendMessageToService(command: MSG.NEXT);
		}

		private void updateCustomConfigString(String customConfigString) {
			var wpsi = WallpaperServiceInfo.getInstance();
			// request to update service only when it's different from previous set
			if (wpsi.getCustomConfigString() != customConfigString) {
				_serviceConnection.sendMessageToService(MSG.CUSTOM_CONFIG, 0, customConfigString);
			}
		}

		private void customConfigStringChanged(object sender, EventArgs e) {
            BPUtil.BPLog("WallpaperInfoUI.customConfigStringChanged");
			String newCustomConfigString = _customConfigStringTextBox.Text;
			// lose focus to be updated
			// self.window?.makeFirstResponder(_themeComboBox);
			updateCustomConfigString(newCustomConfigString);
			_themeComboBox.Focus();
		}

        private void customConfigStringSetButtonClicked(object sender, EventArgs e) {
            BPUtil.BPLog("WallpaperInfoUI.customConfigStringSetButtonClicked");
			WpfCustomControlLib.CustomConfigDialog.openCustomConfigDialog(WallpaperServiceInfo.getInstance().getCustomConfigString(),
				customConfigString => {
				if (customConfigString != null && customConfigString != "") {
					updateCustomConfigString(customConfigString);
					_customConfigStringTextBox.Text = customConfigString;
				}
			});
		}

		private void updateLayoutWithServiceInfo() {
			if (AppUtil.isMyServiceRunning()) {
				_serviceRunning = true;
			} else {
				_serviceRunning = false;
			}

			WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();

            var rootPath = wpsi.getSourceRootPath().Replace(BPUtil.getHomeDirectory(), "~/").Replace("\\", "/");
            var pretty_root_path = Regex.Replace(rootPath, @"(.):/", "/$1/", RegexOptions.IgnoreCase);
			_rootPathTextBox.Text = pretty_root_path;
		
			_intervalTrackBar.Value = wpsi.getInterval();
		
			if (!_serviceRunning) {
				_pauseButton.Image = global::WallpaperInfoApp.Properties.Resources.media_play_pause;
			} else if (wpsi.getPause()) {
				_pauseButton.Image = global::WallpaperInfoApp.Properties.Resources.media_play;
			} else {
				_pauseButton.Image = global::WallpaperInfoApp.Properties.Resources.media_pause;
			}
			_doneLoading = false;
			if (_serviceRunning) {
				_dirButton.Enabled = true;
				_themeComboBox.Enabled = true;
				_themeComboBox.SelectedIndex = wpsi.getTheme().intValue();
				_intervalTrackBar.Enabled = true;
				// update only when it's not being edited
				if (!_customConfigStringTextBox.Focused) {
					_customConfigStringTextBox.Text = wpsi.getCustomConfigString();
				}
				_customConfigStringTextBox.Enabled = true;
				_customConfigStringSetButton.Enabled = true;
				_previousButton.Enabled = true;
				_pauseButton.Enabled = true;
				_nextButton.Enabled = true;
				_startStopButton.Image = global::WallpaperInfoApp.Properties.Resources.media_stop;
			} else {
				_dirButton.Enabled = false;
				_themeComboBox.Enabled = false;
				_intervalTrackBar.Enabled = false;
				_customConfigStringTextBox.Text = "";
				_customConfigStringTextBox.Enabled = false;
				_customConfigStringSetButton.Enabled = false;
				_previousButton.Enabled = false;
				_pauseButton.Enabled = false;
				_nextButton.Enabled = false;
				_startStopButton.Image = global::WallpaperInfoApp.Properties.Resources.power_on;

				// clear list and update
				_lastUsedPathsList.Clear();
				//_lastUsedPathsList.notifyDataSetChanged();
			}
			refreshThumbnail();
			_doneLoading = true;
		}

		// UI rendering callback
		private void wallpaperInfoUpdated(bool isFullUpdate) {
			if (isFullUpdate) {
			// WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
			// makeListViewItem(wpsi.getLastUsedPaths());
			}
			this.Invoke((MethodInvoker)delegate {
				this.updateLayoutWithServiceInfo();
			});
		}

		private void refreshThumbnail() {
			if (_serviceRunning) {
				WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
				if (wpsi.getThumbnail() != null) {
					_thumbnailView.Image = wpsi.getThumbnail();
					return;
				}
			}
			_thumbnailView.Image = global::WallpaperInfoApp.Properties.Resources.thirdwave;
		}

		private void refreshlastUsedPathsListView() {
		}

		public void broadcastReceiver(Bitmap thumbnail, WPath currentWPath, bool pause) {
			//Console.WriteLine("WallpaperInfoUI.broadcastReceiver()");
			if (WallpaperInfoUI._singleUI != null) {
				wallpaperInfoUpdated(false);
			} else {
				// Console.WriteLine("Broadcast is ignored due to no UI.");
			}
		}

		private WallpaperServiceConnection _serviceConnection = null;
	    private bool _serviceRunning = false;
		private bool _doneLoading = false;

		// private static var _filterStringEditingColor:NSColor = NSColor.red // Color.parseColor("#FFFCDC");
		private List<String> _lastUsedPathsList = new List<String>();
		// private static var _selectedRowColor:NSColor = NSColor.black // = Color.parseColor("#FFFC99");

		// Const value property
		public const int _lastUsedPathsRemovalCounnt = 5;

		public static void showOptionWindow(bool isToggle = false) {
			if (_singleUI == null) {
				_singleUI = new WallpaperInfoUI();
				var screen = Screen.PrimaryScreen;
                _singleUI.Location = new Point(screen.WorkingArea.Right - _singleUI.Width - 5,
                                                    screen.WorkingArea.Bottom - _singleUI.Height - 1);
				_singleUI.Show();
			} else if (isToggle) {
				_singleUI.Close();
				_singleUI = null;
				return;
			}
			// trick to bring app to the front most
			_singleUI.TopMost = true;
			_singleUI.Focus();
			_singleUI.TopMost = false;
			_singleUI.BringToFront();
		}

		private void thumbnailViewClicked(object sender, EventArgs e) {
            BPUtil.BPLog("WallpaperInfoUI.thumbnailViewClicked");
			WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
			WPath wpath = wpsi.getcurrentWPath();
			if (BPUtil.shiftKeyPressed()) {
				BPUtil.showImageFile(wpath.path);
			} else {
				BPUtil.showImagePreview(wpath.path);
			}
		}

		private void wallpapaerInfoUIClosed(object sender, FormClosedEventArgs e) {
			_singleUI = null;
		}

		private static WallpaperInfoUI _singleUI = null;
    }
}

File: macosx/WallpaperInfoApp/WallpaperInfoApp/WallpaperInfoUI.swift
//
//  WallpaperInfoWindowController.swift
//  WallpaperInfoApp
//
//  Created by Brad Park on 4/11/20.
//  Copyright © 2020 Brad Park. All rights reserved.
//

import Foundation
import Cocoa

var _wallpaperInfoWindowUI: WallpaperInfoUI?

class WallpaperInfoUI: NSWindowController, NSWindowDelegate {
    
    override func windowDidLoad() {
        super.windowDidLoad()
        // must assign to retain the instance of WallpaperInfoWindowController
        _wallpaperInfoWindowUI = self
		
		if (AppUtil.isMyServiceRunning()) {
			_serviceRunning = true
			_serviceConnection = WallpaperServiceConnection.bindService(WallpaperInfoWindowController: self)
		} else {
			_serviceRunning = false
		}
		_themeComboBox.addItems(withObjectValues: ThemeInfo.getLabels())
        updateLayoutWithServiceInfo()
		self.window?.delegate = self
    }
	
    @IBAction func chooseDirButtonPressed(sender: AnyObject?) {
        BPUtil.BPLog("chooseDirButtonPressed")
		DirectoryChooserDialog.directoryChooserDialog(window: self.window!, completion: { (path) in
			if path != nil {
				self._serviceConnection?.sendMessageToService(command: MSG.SET_ROOT, objectOption: path)
				DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
					self.updateLayoutWithServiceInfo()
				}
			}
		})
    }
	
	@IBAction func themeComboBoxChanged(sender: AnyObject?) {
        BPUtil.BPLog("themeComboBoxChanged")
		let index = _themeComboBox.indexOfSelectedItem
		_serviceConnection?.sendMessageToService(command:MSG.SET_THEME, intOption: index)
	}
	
	@IBAction func saverButtonPressed(sender: AnyObject?) {
        BPUtil.BPLog("saverButtonPressed")
		_serviceConnection?.sendMessageToService(command:MSG.SET_SAVER, intOption: _saverButton.state == NSControl.StateValue.on ? 1 : 0)
	}
	
	@IBAction func intervalSeekbarChanged(sender: AnyObject?) {
        BPUtil.BPLog("intervalSeekbarChanged " + String(_intervalSeekBar.intValue))
		_serviceConnection?.sendMessageToService(command:MSG.SET_INTERVAL, intOption: Int(_intervalSeekBar.intValue))
	}
	
    @IBAction func startStopButtonPressed(sender: AnyObject?) {
        BPUtil.BPLog("startStopButtonPressed")
        if (AppUtil.isMyServiceRunning()) {
            _serviceRunning = true;
        } else {
            _serviceRunning = false;
        }
        
        if (_serviceRunning) {
			WallpaperServiceConnection.stopService()
			_serviceConnection?.unbindService()
			_serviceConnection = nil
        } else {
			_serviceConnection = WallpaperServiceConnection.bindService(WallpaperInfoWindowController: self)
			DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
				WallpaperServiceConnection.startService()
			}
        }
        _serviceRunning = !_serviceRunning;
		DispatchQueue.main.asyncAfter(deadline: .now()) {
			self.updateLayoutWithServiceInfo()
		}
    }
    
    @IBAction func previousButtonPressed(sender: AnyObject?) {
        BPUtil.BPLog("previousButtonPressed")
		_serviceConnection?.sendMessageToService(command: MSG.PREVIOUS)
    }
    
    @IBAction func pauseButtonPressed(sender: AnyObject?) {
        BPUtil.BPLog("pauseButtonPressed")
        let wpsi = WallpaperServiceInfo.getInstance()
		_serviceConnection?.sendMessageToService(command: MSG.PAUSE, intOption: (!wpsi.getPause()) ? 1:0)
		updateLayoutWithServiceInfo()
    }
    
    @IBAction func nextButtonPressed(sender: AnyObject?) {
        BPUtil.BPLog("nextButtonPressed")
		_serviceConnection?.sendMessageToService(command: MSG.NEXT)
    }
	
	private func updateCustomConfigString(customConfigString:String) {
		let wpsi = WallpaperServiceInfo.getInstance()
		// request to update service only when it's different from previous set
		if (wpsi.getCustomConfigString() != customConfigString) {
			_serviceConnection?.sendMessageToService(command: MSG.CUSTOM_CONFIG, objectOption: customConfigString)
		}
	}

	@IBAction func customConfigStringEdited(sender: AnyObject?) {
        BPUtil.BPLog("customConfigStringEdited")
		let newCustomConfigString = _customConfigStringTextField.stringValue
		updateCustomConfigString(customConfigString: newCustomConfigString)
		// lose focus to be updated
		self.window?.makeFirstResponder(_themeComboBox)
		_themeComboBox.selectText(self)
	}

    @IBAction func customConfigButtonPressed(sender: AnyObject?) {
        BPUtil.BPLog("customConfigButtonPressed")
		CustomConfigDialog.openCustomConfigDialog(window: self.window!, completion: { (customConfigString) in
			if let newCustomConfigString = customConfigString {
				self.updateCustomConfigString(customConfigString: newCustomConfigString)
				// update UI with dialog return
				self._customConfigStringTextField.stringValue = newCustomConfigString
			}
		})
    }
    
    private func updateLayoutWithServiceInfo() {
        if (AppUtil.isMyServiceRunning()) {
            _serviceRunning = true;
        } else {
            _serviceRunning = false;
        }
        let wpsi = WallpaperServiceInfo.getInstance()
		let rootPath = wpsi.getSourceRootPath()
        let pretty_root_path = rootPath.replacingOccurrences(of: BPUtil.getHomeDirectory(), with: "~")
        _rootPathTextField.stringValue = pretty_root_path
		_intervalSeekBar.intValue = Int32(wpsi.getInterval())
		if (!_serviceRunning) {
			_pauseButton.image = NSImage(named: "media_play_pause.png")
		} else if (wpsi.getPause()) {
            _pauseButton.image = NSImage(named: "media_play.png")
        } else {
            _pauseButton.image = NSImage(named: "media_pause.png")
        }
        if (_serviceRunning) {
			_dirButton.isEnabled = true
			_themeComboBox.isEnabled = true
			_themeComboBox.selectItem(at: wpsi.getTheme().intValue)
			_saverButton.isEnabled = true
			_saverButton.state = wpsi.getSaver() ? NSControl.StateValue.on : NSControl.StateValue.off
            _intervalSeekBar.isEnabled = true
			// update only when it's not being edited
			if (_customConfigStringTextField.currentEditor() == nil || _customConfigStringTextField.stringValue == "") {
				_customConfigStringTextField.stringValue = wpsi.getCustomConfigString()
			}
			_customConfigStringTextField.isEnabled = true
            _previousButton.isEnabled = true
            _pauseButton.isEnabled = true
            _nextButton.isEnabled = true
            _startStopButton.image = NSImage(named: "media_stop.png")
        } else {
            _dirButton.isEnabled = false
			_themeComboBox.isEnabled = false
			_saverButton.isEnabled = false
			_saverButton.state = NSControl.StateValue.off
            _intervalSeekBar.isEnabled = false
            _customConfigStringTextField.stringValue = ""
            _customConfigStringTextField.isEnabled = false 
            _previousButton.isEnabled = false
            _pauseButton.isEnabled = false
            _nextButton.isEnabled = false
			_startStopButton.image = NSImage(named: "power_on.png")
            // clear list and update
			_lastUsedPathsList.removeAll()
            //_wallpaperAdapter.notifyDataSetChanged();
        }
        refreshThumbnail();
		/*
        _lastUsedPathsListView.invalidateViews();
        _lastUsedPathsListView.refreshDrawableState();*/
    }
	
	// UI rendering callback
	@objc func wallpaperInfoUpdated(isFullUpdate:Bool) {
		if (isFullUpdate) {
			// let wpsi = WallpaperServiceInfo.getInstance();
			// makeListViewItem(wpsi.getLastUsedPaths());
		}
		DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
			self.updateLayoutWithServiceInfo()
		}
	}

	func refreshThumbnail() {
		if (_serviceRunning) {
			let wpsi = WallpaperServiceInfo.getInstance()
			if (wpsi.getThumbnail() != nil) {
				_thumbnailView.image = wpsi.getThumbnail()
                return
			}
		}
        _thumbnailView.image = NSImage(named: "thirdwave.png")
	}
	
	func refreshlastUsedPathsListView() {
/*		WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
		String path = wpsi.getcurrentWPath();
		if (path != null) {
			int index = _lastUsedPathsList.indexOf(path);
			if (index < 0) {
				if (_lastUsedPathsList.size() + _lastUsedPathsRemovalCounnt + 1 > WallpaperService._maxLastUsedPaths) {
					_lastUsedPathsList.removeLast();
				}
				_lastUsedPathsList.addFirst(path);
				index =  0;

			}
			_wallpaperAdapter.setSelectedRow(index);
			_lastUsedPathsListView.smoothScrollToPosition(index);
		}
		_wallpaperAdapter.notifyDataSetChanged();
*/
	}
	
	func broadcastReceiver(thumbnail:NSImage?, currentWPath:WPath?, pause:Bool) {
		// print("WallpaperInfoUI.broadcastReceiver()")
		if (WallpaperInfoUI._singleWindow != nil) {
			wallpaperInfoUpdated(isFullUpdate: false)
		} else {
			// print("Error - Broadcast shouldn't be delivered without UI")
		}
	}
	
    // followings are not needed with init(windowNibName: "nib_name")
	/*
    override var windowNibName: String? { return "WallpaperInfoUI" }
    override var owner: AnyObject? { return self }
	*/
	private var _serviceConnection:WallpaperServiceConnection? = nil
    private var _serviceRunning:Bool = false;

    @IBOutlet weak var _rootPathTextField: NSTextField!
    @IBOutlet weak var _themeComboBox: NSComboBox!
    @IBOutlet weak var _saverButton: NSButton!
    @IBOutlet weak var _intervalSeekBar: NSSlider!
    @IBOutlet weak var _dirButton: NSButton!
    @IBOutlet weak var _pauseButton: NSButton!
    @IBOutlet weak var _startStopButton: NSButton!
    @IBOutlet weak var _previousButton: NSButton!
    @IBOutlet weak var _nextButton: NSButton!
    @IBOutlet weak var _thumbnailView: ThumbnailView!
    @IBOutlet weak var _customConfigStringSetButton: NSButton!
    @IBOutlet weak var _customConfigStringTextField: NSTextField!
    @IBOutlet weak var _lastUsedPathsTableView: NSTableView!

    private static var _filterStringEditingColor:NSColor = NSColor.red // Color.parseColor("#FFFCDC");
	private var _lastUsedPathsList:Array<String> = Array<String>()
    private static var _selectedRowColor:NSColor = NSColor.black // = Color.parseColor("#FFFC99");

    // Const value property
    public static var _lastUsedPathsRemovalCounnt:Int = 5
	
	// Ensure NSWindow Single Instance
    class func showOptionWindow(isToggle:Bool = false) {
		if (_singleWindow == nil) {
			let wallpaperInfoWindowController = WallpaperInfoUI.init(windowNibName: "WallpaperInfoUI")
            if let sf = NSScreen.main?.visibleFrame {
                let wf = wallpaperInfoWindowController.window!.frame
                wallpaperInfoWindowController.window!.setFrame(
                    NSMakeRect(NSWidth(sf) - NSWidth(wf),
							   NSHeight(sf) - NSHeight(wf) + sf.minY,
							   wf.size.width,
							   wf.size.height), display: true)
            }
			wallpaperInfoWindowController.showWindow(nil)
			_singleWindow=wallpaperInfoWindowController.window
		} else if (isToggle) {
            _singleWindow?.close()
            _singleWindow = nil
            return
        }
		// Bring Window to front in either case
		NSApp.activate(ignoringOtherApps: true)
	}
	
	func windowWillClose(_ notification: Notification) {
		_serviceConnection?.unbindService()
		WallpaperInfoUI._singleWindow = nil
	}
	
	static public let _defaultCustomConfigString =
					ThemeInfo._default_custom_root + ";" +
					ThemeInfo._default_custom_allow + ";" +
					ThemeInfo._default_custom_filter
	
	static var _singleWindow:NSWindow? = nil
    
}

class ThumbnailView: NSImageView {
    override func mouseDown(with event: NSEvent) {
		super.mouseDown(with: event)
        BPUtil.BPLog("ThumbnailView mouseDown")
		let wpsi = WallpaperServiceInfo.getInstance()
		if let currentWPath = wpsi.getcurrentWPath() {
            if (BPUtil.optionKeyPressed()) {
				BPUtil.showImageFile(with: currentWPath.path)
                } else {
                BPUtil.openImagePreview(with: currentWPath.path)
            }
		}
    }
}


